### ISSUE TITLE
support for c++ threads and new operator

### ISSUE THREAD

**Issue by adevelopcr** [DarthTon/Blackbone#259]
I've learned much from this project and I'm building my own injector but I have problems with c++ std threads and new

when loaded dll uses c++ std like threads new to allocate memory it crashes immediately

when I use new I get : 
Unhandled exception at (the address of RtlAllocateHeap) (ntdll.dll) in myloader.exe: 0xC0000005: Access violation reading location 0x0000000000000074 (this is always the same ! strange thing)

and when using threads I get crash here : 
83 3C 01 00          cmp         dword ptr [rcx+rax],0 ; rax is zero , rcx is address of read access violation

I'm using visual studio 2015 and can't move now to 2017 so I couldn't build blackbone but I navigate the source code

If blackbone has any solution can you tell me where to find it in the source

**Comment by adevelopcr** [comment]
you may also need this : 
http://www.nynaeve.net/Code/VistaImplicitTls.cpp

**Comment by adevelopcr** [comment]
any help ?

**Comment by DarthTon** [comment]
Give me your test dll.

**Comment by adevelopcr** [comment]
this is the thread test , it crashes when calling DllMain : 
 `int msg(int i) {
    MessageBoxA(0, to_string(i).c_str(), 0, 0);
    return i+5;
}
extern "C" __declspec(dllexport) void testThread() {
    thread(msg, 5).detach();
}
BOOL WINAPI DllMain(HINSTANCE h,DWORD reason, LPVOID reserved ) {
return TRUE;
}`

this is new operator problem : 

`extern "C" __declspec(dllexport) void whatabout() {
    char *sometext = new char[500]; // crash here
    sometext = "this is some text";
    MessageBoxA(0, sometext, sometext, 0);
}
  
int __stdcall DllMain(HINSTANCE h,DWORD reason, LPVOID reserved ) {
    return 1;
}`

if I used VirtualAlloc or HeapAlloc or overloaded new to use  HeapAlloc  it doesn't crash

note : I found that it isn't only thread that crashes but also mutex , condition_variable , future , promis , ....
however std::atomic works  

built these dlls with visual studio 2015 /MT Release

**Comment by adevelopcr** [comment]
I tried RtlAddFunctionTable but didn't work however the function returns tru

        `DWORD size = optHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_EXCEPTION].Size;
	uintptr_t count = (size - sizeof(IMAGE_RUNTIME_FUNCTION_ENTRY)) / 
       sizeof(IMAGE_RUNTIME_FUNCTION_ENTRY);
	RtlAddFunctionTable((PRUNTIME_FUNCTION)pruntime, count,image); // image is the base , pruntime is PIMAGE_RUNTIME_FUNCTION_ENTRY `

**Comment by staticinvocation** [comment]
I have a feeling your CRT is not being initialized.  Are you calling TLS callbacks?  On MSVC2015+ I believe CRT initialization is done with TLS callbacks.  It's possible that the CRT allocator requires initialization via CRT initialization, or it might use TLS to store some information required by the allocator, which would explain why it's crashing only in a new thread.

Also I believe std::atomic works as essentially an x86 `lock` wrapper so it wouldn't be reliant on any CRT initialization.

**Comment by adevelopcr** [comment]
Yes I init the tls I got the relative address of the function that initializes it from ntdll pdb on my system for both 32 an 64 ntdll and it works just fine

**Comment by staticinvocation** [comment]
ntdll is gonna be fine since it's loaded by the system loader, LdrLoadLibrary.  You need to worry about calling TLS callbacks in your own module, and given by your description, it doesn't seem like you are doing this, at least properly.

**Comment by adevelopcr** [comment]
I'm using this function LdrpHandleTlsData with relative address and set breakpoint on this addeess added to ntdll base and visual studio shows that this is the address of LdrpHandleTlsData

I tried it with __declspec(thread) int variable , when I don't use it the variable is 0 in main  but when I used it the variable was correctly set

**Comment by staticinvocation** [comment]
Initializing static TLS is great, but that's not what I said.  You need to call TLS **callbacks** which `LdrpHandleTlsData` does _not_ do.

https://github.com/DarthTon/Blackbone/blob/master/src/BlackBone/ManualMap/MMap.cpp#L1175

**Comment by adevelopcr** [comment]
I didn't know this , I'll try when I get to the compiler
I also found this issue : 
https://github.com/DarthTon/Blackbone/issues/98

**Comment by adevelopcr** [comment]
tls callbacks wasn't the solution

**Comment by DarthTon** [comment]
I still think something is wrong with CRT initialization but can't tell what exactly, not enough info.

**Comment by adevelopcr** [comment]
What info do you need ? and did you try the simple dlls I put above that crash immediately when calling DllMain ?

**Comment by DarthTon** [comment]
You sample works fine on my Win10 but is broken on Win7. At least I can reproduce it now,

**Comment by adevelopcr** [comment]
@DarthTon hmm , this problem also was on windows 7 and nearly the as mine : 
https://github.com/DarthTon/Blackbone/issues/235
I think there is something in windowses earlier than 10

**Comment by DarthTon** [comment]
Yes, root cause is the same as in https://github.com/DarthTon/Blackbone/issues/235. It works on Win10 because `RtlPcToFileHeader` started to use `InvertedFunctionTable` instead of `Ldr.InLoadOrderLinks`. 
So to fix this either add module to `Ldr.InLoadOrderLinks` or hook `RtlPcToFileHeader` and return proper pointer from there for the mapped module.

**Comment by adevelopcr** [comment]
but RtlPcToFileHeader isn't in the import table how can I hook it in the buffer of dll I'm loading in memory I can only hook by faking the address in import table or using some hooking library which requires the target to be already running

sorry I am not good at hooking can you explain more

**Comment by DarthTon** [comment]
Parse export table manually. But the function should be present in the import table of DLL with /MT.

**Comment by adevelopcr** [comment]
and won't this crash the target when it uses the hooked function

**Comment by DarthTon** [comment]
no if you hook it properly

**Comment by adevelopcr** [comment]
sorry for late I didn't use the computer some days ago
now I checked the import table and found that the used function is RtlLookupFunctionEntry not  RtlPcToFileHeader 

it also returns the base of the module containing value but also have third parameter which is PUNWIND_HISTORY_TABLE
I don't know how to get this parameter for my manually mapped module

**Comment by adevelopcr** [comment]
I checked the wrong file !!!
I had MD and MT dll and I checked MD which didn't have RtlPcToFileHeader
but MT dll has it

**Comment by adevelopcr** [comment]
I manual mapped the dll in the current process (not remote process) and hooked the function with my custom one and it worked very well !!!

there is something I'm missing and I hope you help me with it
it's about remote hooking

the technique I used to hook the function is to hook while iat resolving and the hook function used a global uintptr_t which holds the base address of the manual mapped image

now if I hook in a remote process I should write the hook function and Loader function and pass the address of hook function as a parameter to Loader , the problem is : hook function won't know the base address of the mapped image even I wrote the address in the target process as hook function has only the address of the global variable in my process
I'm also going to manage tls callbacks with the same technique
can you help me get around this ?

**Comment by DarthTon** [comment]
Make image base an exportable symbol from your dll, set value remotely and use it in hook.

**Comment by adevelopcr** [comment]
I used a hook function , wrote it in the target exe , patched it there with the Loader function I wrote in the target 
the patch set the address of RtlPcFileToHeader and the image base in the target and hooked the function during resolving imports and now it works very well

do I need to hook CreateThread also to manage tls callbacks or LdrpHandleTlsData wil do the job ?
